usr.bin/uucp/uuclean/uuclean.8
.\" Copyright (c) 1983, 1991 Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)uuclean.8	6.7 (Berkeley) 9/13/91
.\"
.Dd September 13, 1991
.Dt UUCLEAN 8
.Os BSD 4.2
.Sh NAME
.Nm uuclean
.Nd uucp spool directory clean-up
.Sh SYNOPSIS
.Nm uuclean
.Op Fl d Ns Ar subdirectory
.Op Fl m
.Op Fl n Ns Ar time
.Op Fl p Ns Ar pre
.Sh DESCRIPTION
.Xr Uuclean
will scan the spool directory for files with the
specified prefix and delete all those which are
older than the specified number of hours.
.Pp
The following options are available.
.Bl -tag -width Ds
.It Fl d Ns Ar subdirectory
Only the specified subdirectory will be cleaned.
.It Fl m
Send mail to the owner of the file
when it is deleted.
.It Fl n Ns Ar time
Files whose age is more than
.Ar time
hours will be deleted if the prefix
test is satisfied.
(default time is 72 hours)
.It Fl p Ns Ar pre
Scan for files with
.Ar pre
as the file prefix.
Up to 10
.Fl p
arguments may be specified.
.El
.Pp
This program will typically be run daily by
.Xr cron 8 .
.Sh FILES
.Bl -tag -width /var/spool/uucp
.It Pa /var/spool/uucp
Spool directory
.El
.Sh SEE ALSO
.Xr uucp 1 ,
.Xr uux 1 ,
.Xr uucico 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .


usr.bin/uucp/uuclean/uuclean.c
/*-
 * Copyright (c) 1985 The Regents of the University of California.
 * All rights reserved.
 *
 * This module is believed to contain source code proprietary to AT&T.
 * Use and redistribution is subject to the Berkeley Software License
 * Agreement and your Software Agreement with AT&T (Western Electric).
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1985 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)uuclean.c	5.9 (Berkeley) 4/24/91";
#endif /* not lint */

#include <signal.h>
#include "uucp.h"
#include <pwd.h>
#include <sys/stat.h>
#ifdef	NDIR
#include "ndir.h"
#else
#include <sys/dir.h>
#endif

/*
 *
 *	this program will search through the spool
 *	directory (Spool) and delete all files with a requested
 *	prefix which are older than (nomtime) seconds.
 *	If the -m option is set, the program will try to
 *	send mail to the usid of the file.
 *
 *	options:
 *		-m  -  send mail for deleted file
 *		-d  -  directory to clean
 *		-n  -  time to age files before delete (in hours)
 *		-p  -  prefix for search
 *		-x  -  turn on debug outputs
 *	exit status:
 *		0  -  normal return
 *		1  -  can not read directory
 */

#define NOMTIME 72	/* hours to age files before deletion */

int checkprefix = 0;
struct timeb Now;

main(argc, argv)
char *argv[];
{
	register DIR *dirp;
	register struct direct *dentp;
	time_t nomtime, ptime;
	struct stat stbuf;
	int mflg = 0;

	strcpy(Progname, "uuclean");
	uucpname(Myname);
	nomtime = NOMTIME * (time_t)3600;

	while (argc>1 && argv[1][0] == '-') {
		switch (argv[1][1]) {
		case 'd':
			Spool = &argv[1][2];
			break;
		case 'm':
			mflg = 1;
			break;
		case 'n':
			nomtime = atoi(&argv[1][2]) * (time_t)3600;
			break;
		case 'p':
			checkprefix = 1;
			if (&argv[1][2] != '\0')
				stpre(&argv[1][2]);
			break;
		case 'x':
			chkdebug();
			Debug = atoi(&argv[1][2]);
			if (Debug <= 0)
				Debug = 1;
			break;
		default:
			printf("unknown flag %s\n", argv[1]); break;
		}
		--argc;  argv++;
	}

	DEBUG(4, "DEBUG# %s\n", "START");
	if (chdir(Spool) < 0) {	/* NO subdirs in uuclean!  rti!trt */
		printf("%s directory inaccessible\n", Spool);
		exit(1);
	}

	if ((dirp = opendir(Spool)) == NULL) {
		printf("%s directory unreadable\n", Spool);
		exit(1);
	}

	time(&ptime);
	while (dentp = readdir(dirp)) {
		if (checkprefix && !chkpre(dentp->d_name))
			continue;

		if (stat(dentp->d_name, &stbuf) == -1) {
			DEBUG(4, "stat on %s failed\n", dentp->d_name);
			continue;
		}


		if ((stbuf.st_mode & S_IFMT) == S_IFDIR)
			continue;
		if ((ptime - stbuf.st_mtime) < nomtime)
			continue;
		if (dentp->d_name[0] == CMDPRE)
			notfyuser(dentp->d_name);
		DEBUG(4, "unlink file %s\n", dentp->d_name);
		unlink(dentp->d_name);
		if (mflg)
			sdmail(dentp->d_name, stbuf.st_uid);
	}

	closedir(dirp);
	exit(0);
}


#define MAXPRE 10
char Pre[MAXPRE][NAMESIZE];
int Npre = 0;
/***
 *	chkpre(file)	check for prefix
 *	char *file;
 *
 *	return codes:
 *		0  -  not prefix
 *		1  -  is prefix
 */

chkpre(file)
char *file;
{
	int i;

	for (i = 0; i < Npre; i++) {
		if (prefix(Pre[i], file))
			return(1);
		}
	return(0);
}

/***
 *	stpre(p)	store prefix
 *	char *p;
 *
 *	return codes:  none
 */

stpre(p)
char *p;
{
	if (Npre < MAXPRE - 2)
		strcpy(Pre[Npre++], p);
	return;
}

/***
 *	notfyuser(file)	- notfiy requestor of deleted requres
 *
 *	return code - none
 */

notfyuser(file)
char *file;
{
	FILE *fp;
	int numrq;
	char frqst[100], lrqst[100];
	char msg[BUFSIZ];
	char *args[10];

	if ((fp = fopen(file, "r")) == NULL)
		return;
	if (fgets(frqst, 100, fp) == NULL) {
		fclose(fp);
		return;
	}
	numrq = 1;
	while (fgets(lrqst, 100, fp))
		numrq++;
	fclose(fp);
	sprintf(msg,
	  "File %s delete. \nCould not contact remote. \n%d requests deleted.\n", file, numrq);
	if (numrq == 1) {
		strcat(msg, "REQUEST: ");
		strcat(msg, frqst);
	} else {
		strcat(msg, "FIRST REQUEST: ");
		strcat(msg, frqst);
		strcat(msg, "\nLAST REQUEST: ");
		strcat(msg, lrqst);
	}
	getargs(frqst, args, 10);
	mailst(args[3], msg, CNULL);
}


/***
 *	sdmail(file, uid)
 *
 *	sdmail  -  this routine will determine the owner
 *	of the file (file), create a message string and
 *	call "mailst" to send the cleanup message.
 *	This is only implemented for local system
 *	mail at this time.
 */

sdmail(file, uid)
char *file;
{
	static struct passwd *pwd;
	char mstr[40];

	sprintf(mstr, "uuclean deleted file %s\n", file);
	if (pwd != NULL && pwd->pw_uid == uid) {
		mailst(pwd->pw_name, mstr, CNULL);
		return;
	}

	setpwent();
	if ((pwd = getpwuid(uid)) != NULL)
		mailst(pwd->pw_name, mstr, CNULL);
}

cleanup(code)
int code;
{
	exit(code);
}
